### PR TITLE
Added testing for the backend modules

### LINKED ISSUE #56: Write tests for backend modules
Write tests in ```pytest``` for all backend functions

### PR DESCRIPTION
**Pull request by andrew-fennell**
These simple tests will help a reviewer see if the PR is breaking something.
I'm going to add some more tests and maybe some linting checks in the future.

The tests run on 3 operating systems: 
- windows-latest
- macos-latest
- ubuntu-latest

On each machine, it does the tests on python versions:
- 3.7
- 3.8
- 3.9

Closes #56

### CODE REVIEW

File: CogNative/backend/modules/STT.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import speech_recognition as sr
 
-from modules import languages
+from .languages import available_languages
```
Comment (andrew-fennell) [inline CogNative/backend/modules/STT.py:3]: Is this change okay? With this change, the importing doesn't working when running the script independently from the ```backend/modules``` directory, but it works when importing this module from ```main.py``` (or any other parent directory).

The previous ```from modules import languages``` worked when running the script from ```backend/modules```, but did not work when accessing it from a parent directory.

File: CogNative/backend/modules/translation.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 from googletrans import Translator
 
-from modules import languages
+from .languages import available_languages
```
Comment (andrew-fennell) [inline CogNative/backend/modules/translation.py:3]: See previous comment.

Comment (andrew-fennell) [inline CogNative/backend/modules/translation.py:3]: See previous comment.

File: requirements.txt [hunk 3]
```diff
@@ -13,6 +13,7 @@ inflect==5.3.0
 ipython>=7.31.1
 librosa==0.8.1
 matplotlib==3.5.1
+nltk==3.7
```
Comment (andrew-fennell) [inline requirements.txt:16]: This import is being used in testing to determine the similarity between two sentences.

Comment (andrew-fennell) [inline requirements.txt:16]: This import is being used in testing to determine the similarity between two sentences.

File: tests/backend/modules/test_STT.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                   "he's been in office he's been an absolute"
+                   " power for 20 years what is left what "
+                   "you know what's in the soul of a man he "
+                   "wants to be a hero and he thinks this is "
+                   "the way to do it I think people know "
+                   "this is history major stuff")
+    
+    # Check if synthesized text vs actual text is greater than 90%
+    smoothie = SmoothingFunction().method4
+    assert bleu([actual_text], text, smoothing_function=smoothie) > 0.9
```
Comment (andrew-fennell) [inline tests/backend/modules/test_STT.py:31]: Checking the similarity between *actual* text and *transcribed* text.

Comment (andrew-fennell) [inline tests/backend/modules/test_STT.py:31]: Checking the similarity between *actual* text and *transcribed* text.

File: tests/backend/modules/test_STT.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                    " who used the resources they had "
+                    "available to make life work")
+    
+    # Check if synthesized text vs actual text is greater than 90%
+    assert bleu([actual_text2], text2, smoothing_function=smoothie) > 0.9
+
+    # Check if get_transcriptions() is storing the correct
+    # correct text with the source audio file
+    assert s.get_transcriptions()[example_path] == text
+    assert s.get_transcriptions()[example_path2] == text2
```
Comment (andrew-fennell) [inline tests/backend/modules/test_STT.py:49]: ```STT.get_transcriptions``` returns a ```dict``` with path's associated with text transcribed from the audio at that path. These assertions are checking that the text is associated with the correct path.

Comment (andrew-fennell) [inline tests/backend/modules/test_STT.py:49]: ```STT.get_transcriptions``` returns a ```dict``` with path's associated with text transcribed from the audio at that path. These assertions are checking that the text is associated with the correct path.

File: tests/backend/modules/test_translation.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    "de paja.")
+
+def test_invalid_src_language():
+    src_lang = 'elvish' # Unsupported language
+    dest_lang = 'english'
+
+    tr = translation()
+
+    with pytest.raises(Exception):
+        translated_text = tr.translate(english_text, src_lang, dest_lang)
```
Comment (andrew-fennell) [inline tests/backend/modules/test_translation.py:55]: Checks that an error is raised when an invalid language is given.

Comment (andrew-fennell) [inline tests/backend/modules/test_translation.py:55]: Checks that an error is raised when an invalid language is given.

File: tests/backend/modules/test_translation.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    dest_lang = 'spanish'
+
+    tr = translation()
+    translated_text = tr.translate_to(english_text, dest_lang)
+    
+    # Check if synthesized spanish_text vs translated_text 
+    # is greater than 80%
+    smoothie = SmoothingFunction().method4
+    assert bleu([spanish_text], translated_text,
+                smoothing_function=smoothie) > 0.7
```
Comment (andrew-fennell) [inline tests/backend/modules/test_translation.py:77]: Assertions ensures the translated Spanish text is very similar to the actual Spanish text.

Comment (andrew-fennell) [inline tests/backend/modules/test_translation.py:77]: Assertions ensures the translated Spanish text is very similar to the actual Spanish text.

File: tests/backend/modules/test_translation.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    dest_lang = 'english'
+
+    tr = translation()
+    translated_text = tr.translate(spanish_text, src_lang, dest_lang)
+    
+    # Check if synthesized spanish_text vs translated_text 
+    # is greater than 80%
+    smoothie = SmoothingFunction().method4
+    assert bleu([english_text], translated_text,
+                smoothing_function=smoothie) > 0.7
```
Comment (andrew-fennell) [inline tests/backend/modules/test_translation.py:103]: Testing English destination text from Spanish source text and testing similarity between actual English text.

Comment (andrew-fennell) [inline tests/backend/modules/test_translation.py:103]: Testing English destination text from Spanish source text and testing similarity between actual English text.

File: tests/backend/modules/test_translation.py (hunk 7 above)
Comment (andrew-fennell) [reply tests/backend/modules/test_translation.py:77]: This function ```translate_to()``` also does not require a source language. This test also checks that the function is able to identify the source language.

Comment (andrew-fennell) [reply tests/backend/modules/test_translation.py:77]: This function ```translate_to()``` also does not require a source language. This test also checks that the function is able to identify the source language.

File: CogNative/backend/modules/STT.py [hunk 9]
```diff
@@ -13,11 +13,11 @@ def __init__(self, source_language="english"):
         engine -- Engine used to perform speech recognition
         """
 
-        if source_language not in languages.available_languages.keys():
+        if source_language not in available_languages.keys():
```
Comment (andrew-fennell) [inline CogNative/backend/modules/STT.py:16]: This change is due to the change in the import.

Comment (andrew-fennell) [inline CogNative/backend/modules/STT.py:16]: This change is due to the change in the import.

**Review by andrew-fennell — COMMENTED** [review]
These comments will help a reviewer look at the changes and the added tests.

**Comment by Aref-Sadeghi-Gh** [comment]
Looks Good

**Review by Aref-Sadeghi-Gh — COMMENTED** [review]
All good

**Review by Aref-Sadeghi-Gh — APPROVED** [review]
All good. Can merge

**Pull request merged**